### PR TITLE
[Bug] Add Azure-hosted Dalle Image Generation Support

### LINKED ISSUE #2510: [Bug]: Dalle-Critic not working
### Describe the bug

Followed the notebook [https://github.com/microsoft/autogen/blob/main/notebook/agentchat_image_generation_capability.ipynb](url), but getting the following response:
image

Code:
```python
import autogen

config_list_gpt4 = autogen.config_list_from_json(
    "OAI_CONFIG_LIST",
    filter_dict={
        "model": ["wag-gpt4-128k"],
    },
)

config_list_gpt4_vision = autogen.config_list_from_json(
    "OAI_CONFIG_LIST",
    filter_dict={
        "model": ["gpt4-vision"],
    },
)

config_list_dalle = autogen.config_list_from_json(
    "OAI_CONFIG_LIST",
    filter_dict={
        "model": ["dall-e-3"],
    },
)

gpt_config = {
    "cache_seed": 42,  # change the cache_seed for different trials
    "temperature": 0.7,
    "config_list": config_list_gpt4,
    "timeout": 300,
}

gpt_vision_config = {
    "cache_seed": 42,  # change the cache_seed for different trials
    "temperature": 0.7,
    "config_list": config_list_gpt4_vision,
    "timeout": 300,
}

dalle_config = {
    "cache_seed": 42,  # change the cache_seed for different trials
    "temperature": 0.7,
    "config_list": config_list_dalle,
    "timeout": 300,
}

def _is_termination_message(msg) -> bool:
    # Detects if we should terminate the conversation
    if isinstance(msg.get("content"), str):
        return msg["content"].rstrip().endswith("TERMINATE")
    elif isinstance(msg.get("content"), list):
        for content in msg["content"]:
            if isinstance(content, dict) and "text" in content:
                return content["text"].rstrip().endswith("TERMINATE")
    return False


def critic_agent() -> autogen.ConversableAgent:
    return autogen.ConversableAgent(
        name="critic",
        llm_config=gpt_vision_config,
        system_message=CRITIC_SYSTEM_MESSAGE,
        max_consecutive_auto_reply=3,
        human_input_mode="NEVER",
        is_termination_msg=lambda msg: _is_termination_message(msg),
    )


def image_generator_agent() -> autogen.ConversableAgent:
    # Create the agent
    agent = autogen.ConversableAgent(
        name="dalle",
        llm_config=gpt_vision_config,
        max_consecutive_auto_reply=3,
        human_input_mode="NEVER",
        is_termination_msg=lambda msg: _is_termination_message(msg),
    )

    # Add image generation ability to the agent
    dalle_gen = generate_images.DalleImageGenerator(llm_config=dalle_config)
    image_gen_capability = generate_images.ImageGeneration(
        image_generator=dalle_gen, text_analyzer_llm_config=gpt_config
    )

    image_gen_capability.add_to_agent(agent)
    return agent

def extract_images(sender: autogen.ConversableAgent, recipient: autogen.ConversableAgent) -> Image:
    images = []
    all_messages = sender.chat_messages[recipient]

    for message in reversed(all_messages):
        # The GPT-4V format, where the content is an array of data
        contents = message.get("content", [])
        for content in contents:
            if isinstance(content, str):
                continue
            if content.get("type", "") == "image_url":
                img_data = content["image_url"]["url"]
                images.append(img_utils.get_pil_image(img_data))

    if not images:
        raise ValueError("No image data found in messages.")

    return images

###################################################

dalle = image_generator_agent()
critic = critic_agent()

img_prompt = "robot"

result = dalle.initiate_chat(critic, message=img_prompt)
```

### Steps to reproduce

_No response_

### Model Used

_No response_

### Expected Behavior

_No response_

### Screenshots and logs

_No response_

### Additional Information

```
# Name                    Version                   Build  Channel
aiohttp                   3.9.5                    pypi_0    pypi
aiosignal                 1.3.1                    pypi_0    pypi
annotated-types           0.6.0                    pypi_0    pypi
anyio                     4.3.0                    pypi_0    pypi
appnope                   0.1.4              pyhd8ed1ab_0    conda-forge
asgiref                   3.8.1                    pypi_0    pypi
asttokens                 2.4.1              pyhd8ed1ab_0    conda-forge
attrs                     23.2.0                   pypi_0    pypi
azure-core                1.30.1                   pypi_0    pypi
azure-identity            1.16.0                   pypi_0    pypi
backoff                   2.2.1                    pypi_0    pypi
bcrypt                    4.1.2                    pypi_0    pypi
beautifulsoup4            4.12.3                   pypi_0    pypi
build                     1.2.1                    pypi_0    pypi
bzip2                     1.0.8                h80987f9_5  
ca-certificates           2024.2.2             hf0a4a13_0    conda-forge
cachetools                5.3.3                    pypi_0    pypi
certifi                   2024.2.2                 pypi_0    pypi
cffi                      1.16.0                   pypi_0    pypi
chardet                   5.2.0                    pypi_0    pypi
charset-normalizer        3.2.0                    pypi_0    pypi
chroma-hnswlib            0.7.3                    pypi_0    pypi
chromadb                  0.4.24                   pypi_0    pypi
click                     8.1.7                    pypi_0    pypi
coloredlogs               15.0.1                   pypi_0    pypi
comm                      0.2.2              pyhd8ed1ab_0    conda-forge
contourpy                 1.2.1                    pypi_0    pypi
cryptography              42.0.5                   pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
dataclasses-json          0.6.4                    pypi_0    pypi
dataclasses-json-speakeasy 0.5.11                   pypi_0    pypi
debugpy                   1.6.7           py311h313beb8_0  
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
deprecated                1.2.14                   pypi_0    pypi
dirtyjson                 1.0.8                    pypi_0    pypi
diskcache                 5.6.3                    pypi_0    pypi
distro                    1.9.0                    pypi_0    pypi
docker                    7.0.0                    pypi_0    pypi
emoji                     2.11.0                   pypi_0    pypi
exceptiongroup            1.2.0              pyhd8ed1ab_2    conda-forge
executing                 2.0.1              pyhd8ed1ab_0    conda-forge
fastapi                   0.110.1                  pypi_0    pypi
filelock                  3.13.4                   pypi_0    pypi
filetype                  1.2.0                    pypi_0    pypi
flaml                     2.1.2                    pypi_0    pypi
flatbuffers               24.3.25                  pypi_0    pypi
fonttools                 4.51.0                   pypi_0    pypi
frozenlist                1.4.1                    pypi_0    pypi
fsspec                    2024.3.1                 pypi_0    pypi
google-auth               2.29.0                   pypi_0    pypi
googleapis-common-protos  1.63.0                   pypi_0    pypi
greenlet                  3.0.3                    pypi_0    pypi
grpcio                    1.62.1                   pypi_0    pypi
h11                       0.14.0                   pypi_0    pypi
httpcore                  1.0.5                    pypi_0    pypi
httptools                 0.6.1                    pypi_0    pypi
httpx                     0.27.0                   pypi_0    pypi
huggingface-hub           0.22.2                   pypi_0    pypi
humanfriendly             10.0                     pypi_0    pypi
idna                      3.7                      pypi_0    pypi
importlib-metadata        7.0.0                    pypi_0    pypi
importlib-resources       6.4.0                    pypi_0    pypi
importlib_metadata        7.1.0                hd8ed1ab_0    conda-forge
ipykernel                 6.29.3             pyh3cd1d5f_0    conda-forge
ipython                   8.22.2             pyh707e725_0    conda-forge
jedi                      0.19.1             pyhd8ed1ab_0    conda-forge
jinja2                    3.1.3                    pypi_0    pypi
joblib                    1.4.0                    pypi_0    pypi
jsonpatch                 1.33                     pypi_0    pypi
jsonpath-python           1.0.6                    pypi_0    pypi
jsonpointer               2.4                      pypi_0    pypi
jupyter_client            8.6.1              pyhd8ed1ab_0    conda-forge
jupyter_core              5.5.0           py311hca03da5_0  
kiwisolver                1.4.5                    pypi_0    pypi
kubernetes                29.0.0                   pypi_0    pypi
langchain                 0.1.16                   pypi_0    pypi
langchain-community       0.0.33                   pypi_0    pypi
langchain-core            0.1.44                   pypi_0    pypi
langchain-text-splitters  0.0.1                    pypi_0    pypi
langdetect                1.0.9                    pypi_0    pypi
langsmith                 0.1.49                   pypi_0    pypi
libcxx                    16.0.6               h4653b0c_0    conda-forge
libffi                    3.4.4                hca03da5_0  
libsodium                 1.0.18               h27ca646_1    conda-forge
llama-index               0.10.30                  pypi_0    pypi
llama-index-agent-openai  0.2.2                    pypi_0    pypi
llama-index-cli           0.1.12                   pypi_0    pypi
llama-index-core          0.10.30                  pypi_0    pypi
llama-index-embeddings-azure-openai 0.1.7                    pypi_0    pypi
llama-index-embeddings-openai 0.1.7                    pypi_0    pypi
llama-index-indices-managed-llama-cloud 0.1.5                    pypi_0    pypi
llama-index-legacy        0.9.48                   pypi_0    pypi
llama-index-llms-azure-openai 0.1.6                    pypi_0    pypi
llama-index-llms-openai   0.1.15                   pypi_0    pypi
llama-index-multi-modal-llms-openai 0.1.5                    pypi_0    pypi
llama-index-program-openai 0.1.5                    pypi_0    pypi
llama-index-question-gen-openai 0.1.3                    pypi_0    pypi
llama-index-readers-file  0.1.19                   pypi_0    pypi
llama-index-readers-llama-parse 0.1.4                    pypi_0    pypi
llama-parse               0.4.1                    pypi_0    pypi
llamaindex-py-client      0.1.18                   pypi_0    pypi
lxml                      5.2.1                    pypi_0    pypi
markdown-it-py            3.0.0                    pypi_0    pypi
markdownify               0.12.1                   pypi_0    pypi
markupsafe                2.1.5                    pypi_0    pypi
marshmallow               3.21.1                   pypi_0    pypi
matplotlib                3.8.4                    pypi_0    pypi
matplotlib-inline         0.1.7              pyhd8ed1ab_0    conda-forge
mdurl                     0.1.2                    pypi_0    pypi
mmh3                      4.1.0                    pypi_0    pypi
monotonic                 1.6                      pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
msal                      1.28.0                   pypi_0    pypi
msal-extensions           1.1.0                    pypi_0    pypi
multidict                 6.0.5                    pypi_0    pypi
mypy-extensions           1.0.0                    pypi_0    pypi
ncurses                   6.4                  h313beb8_0  
nest-asyncio              1.6.0              pyhd8ed1ab_0    conda-forge
networkx                  3.3                      pypi_0    pypi
nltk                      3.8.1                    pypi_0    pypi
numpy                     1.26.4                   pypi_0    pypi
oauthlib                  3.2.2                    pypi_0    pypi
onnxruntime               1.17.3                   pypi_0    pypi
openai                    1.21.2                   pypi_0    pypi
openssl                   1.1.1w               h53f4e23_0    conda-forge
opentelemetry-api         1.24.0                   pypi_0    pypi
opentelemetry-exporter-otlp-proto-common 1.24.0                   pypi_0    pypi
opentelemetry-exporter-otlp-proto-grpc 1.24.0                   pypi_0    pypi
opentelemetry-instrumentation 0.45b0                   pypi_0    pypi
opentelemetry-instrumentation-asgi 0.45b0                   pypi_0    pypi
opentelemetry-instrumentation-fastapi 0.45b0                   pypi_0    pypi
opentelemetry-proto       1.24.0                   pypi_0    pypi
opentelemetry-sdk         1.24.0                   pypi_0    pypi
opentelemetry-semantic-conventions 0.45b0                   pypi_0    pypi
opentelemetry-util-http   0.45b0                   pypi_0    pypi
orjson                    3.10.1                   pypi_0    pypi
overrides                 7.7.0                    pypi_0    pypi
packaging                 23.2                     pypi_0    pypi
pandas                    2.2.2                    pypi_0    pypi
parso                     0.8.4              pyhd8ed1ab_0    conda-forge
pexpect                   4.9.0              pyhd8ed1ab_0    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    10.3.0                   pypi_0    pypi
pip                       23.3.1          py311hca03da5_0  
platformdirs              4.2.0              pyhd8ed1ab_0    conda-forge
portalocker               2.8.2                    pypi_0    pypi
posthog                   3.5.0                    pypi_0    pypi
prompt-toolkit            3.0.42             pyha770c72_0    conda-forge
protobuf                  4.25.3                   pypi_0    pypi
psutil                    5.9.0           py311h80987f9_0  
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pulsar-client             3.5.0                    pypi_0    pypi
pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge
pyasn1                    0.6.0                    pypi_0    pypi
pyasn1-modules            0.4.0                    pypi_0    pypi
pyautogen                 0.2.25                   pypi_0    pypi
pycparser                 2.22                     pypi_0    pypi
pydantic                  2.7.0                    pypi_0    pypi
pydantic-core             2.18.1                   pypi_0    pypi
pygments                  2.17.2             pyhd8ed1ab_0    conda-forge
pyjwt                     2.8.0                    pypi_0    pypi
pyparsing                 3.1.2                    pypi_0    pypi
pypdf                     4.2.0                    pypi_0    pypi
pypika                    0.48.9                   pypi_0    pypi
pyproject-hooks           1.0.0                    pypi_0    pypi
python                    3.11.0               hc0d8a6c_3  
python-dateutil           2.9.0.post0              pypi_0    pypi
python-dotenv             1.0.1                    pypi_0    pypi
python-iso639             2024.2.7                 pypi_0    pypi
python-magic              0.4.27                   pypi_0    pypi
pytz                      2024.1                   pypi_0    pypi
pyyaml                    6.0.1                    pypi_0    pypi
pyzmq                     25.1.2          py311h313beb8_0  
rapidfuzz                 3.8.1                    pypi_0    pypi
readline                  8.2                  h1a28f6b_0  
regex                     2024.4.16                pypi_0    pypi
replicate                 0.25.2                   pypi_0    pypi
requests                  2.31.0                   pypi_0    pypi
requests-oauthlib         2.0.0                    pypi_0    pypi
rich                      13.7.1                   pypi_0    pypi
rsa                       4.9                      pypi_0    pypi
safetensors               0.4.3                    pypi_0    pypi
scikit-learn              1.4.2                    pypi_0    pypi
scipy                     1.13.0                   pypi_0    pypi
sentence-transformers     2.7.0                    pypi_0    pypi
setuptools                68.2.2          py311hca03da5_0  
shellingham               1.5.4                    pypi_0    pypi
six                       1.16.0             pyh6c4a22f_0    conda-forge
sniffio                   1.3.1                    pypi_0    pypi
soupsieve                 2.5                      pypi_0    pypi
sqlalchemy                2.0.29                   pypi_0    pypi
sqlite                    3.41.2               h80987f9_0  
stack_data                0.6.2              pyhd8ed1ab_0    conda-forge
starlette                 0.37.2                   pypi_0    pypi
striprtf                  0.0.26                   pypi_0    pypi
sympy                     1.12                     pypi_0    pypi
tabulate                  0.9.0                    pypi_0    pypi
tenacity                  8.2.3                    pypi_0    pypi
termcolor                 2.4.0                    pypi_0    pypi
threadpoolctl             3.4.0                    pypi_0    pypi
tiktoken                  0.6.0                    pypi_0    pypi
tk                        8.6.12               hb8d0fd4_0  
tokenizers                0.19.1                   pypi_0    pypi
torch                     2.2.2                    pypi_0    pypi
tornado                   6.3.3           py311h80987f9_0  
tqdm                      4.66.2                   pypi_0    pypi
traitlets                 5.14.2             pyhd8ed1ab_0    conda-forge
transformers              4.40.0                   pypi_0    pypi
typer                     0.12.3                   pypi_0    pypi
typing-inspect            0.9.0                    pypi_0    pypi
typing_extensions         4.11.0             pyha770c72_0    conda-forge
tzdata                    2024.1                   pypi_0    pypi
unstructured              0.13.2                   pypi_0    pypi
unstructured-client       0.18.0                   pypi_0    pypi
urllib3                   1.26.18                  pypi_0    pypi
uvicorn                   0.29.0                   pypi_0    pypi
uvloop                    0.19.0                   pypi_0    pypi
watchfiles                0.21.0                   pypi_0    pypi
wcwidth                   0.2.13             pyhd8ed1ab_0    conda-forge
websocket-client          1.7.0                    pypi_0    pypi
websockets                12.0                     pypi_0    pypi
wheel                     0.41.2          py311hca03da5_0  
wrapt                     1.16.0                   pypi_0    pypi
xz                        5.4.6                h80987f9_0  
yarl                      1.9.4                    pypi_0    pypi
zeromq                    4.3.5                hebf3989_1    conda-forge
zipp                      3.18.1                   pypi_0    pypi
```

### PR DESCRIPTION
**Pull request by WaelKarkoub**
This PR adds support for Dalle image generation hosted on Azure.

## Related issue number

Closes #2510 

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by ekzhu — APPROVED** [review]

**Comment by ekzhu** [comment]
The failing compressible agent tests is not caused by this PR.

File: autogen/agentchat/contrib/capabilities/generate_images.py [hunk 1]
```diff
@@ -287,5 +299,9 @@ def _validate_resolution_format(resolution: str):
 
 
 def _validate_dalle_model(model: str):
-    if model not in ["dall-e-3", "dall-e-2"]:
-        raise ValueError(f"Invalid DALL-E model: {model}. Must be 'dall-e-3' or 'dall-e-2'")
+    if model not in VALID_DALLE_MODELS:
```
Comment (sonichi) [inline autogen/agentchat/contrib/capabilities/generate_images.py:302]: For Azure models, they are not necessarily named as one of VALID_DALLE_MODELS. They can be named as anything.

File: autogen/agentchat/contrib/capabilities/generate_images.py [hunk 2]
```diff
@@ -108,6 +114,12 @@ def cache_key(self, prompt: str) -> str:
         keys = (prompt, self._model, self._resolution, self._quality, self._num_images)
         return ",".join([str(k) for k in keys])
 
+    def _dalle_client_factory(self, dalle_config: Dict) -> Union[OpenAI, AzureOpenAI]:
+        if dalle_config.get("api_type") == "azure":
+            return AzureOpenAI(api_key=dalle_config["api_key"])
```
Comment (sonichi) [inline autogen/agentchat/contrib/capabilities/generate_images.py:119]: AzureOpenAI requires more fields than api_key.

File: autogen/agentchat/contrib/capabilities/generate_images.py [hunk 3]
```diff
@@ -108,6 +114,12 @@ def cache_key(self, prompt: str) -> str:
...
+        else:
+            return OpenAI(api_key=dalle_config["api_key"])
```
Comment (sonichi) [inline autogen/agentchat/contrib/capabilities/generate_images.py:121]: For OpenAI, I'm not sure if more keys are possible. Why do we pass the `api_key` only here?

**Review by BeibinLi — COMMENTED** [review]
Good PR. My biggest concern is mentioned by Chi. Azure endpoints allow users to rename their model, which is not necessarily named as one of VALID_DALLE_MODELS.

**Comment by WaelKarkoub** [comment]
@sonichi @BeibinLi thank you for the reviews! Initially, I presumed that OpenAI and AzureOpenAI were identical. I’ve been attempting to get access to Azure OpenAI APIs, but it seems like I need a corporate account/email. As I’m unable to test these changes locally I’ll keep this PR in draft until I manage to access Azure OpenA.

**Comment by breakstring** [comment]
> @sonichi @BeibinLi thank you for the reviews! Initially, I assumed that OpenAI and AzureOpenAI were identical. I’ve been attempting to get access to Azure OpenAI APIs, but it seems like I need a corporate account/email. As I’m unable to test these changes locally I’ll keep this PR in draft until I manage to access Azure OpenA.

Hi @WaelKarkoub , any update on this PR.
I just encountered the same issue; I can't use Azure's DALL-E 3 to generate images.
╮(╯▽╰)╭

**Comment by rysweet** [comment]
This PR is against AutoGen 0.2. AutoGen 0.2 has been moved to the 0.2 branch. Please rebase your PR on the 0.2 branch or update it to work with the new AutoGen 0.4 that is now in main.

**Comment by rysweet** [comment]
Hi @WaelKarkoub  - we rebased this PR for you. I think it would not take much to get it over the finish line but it's getting stale. Have a look at giving it an update, otherwise we may close it in a couple of weeks and you can reopen if you get the itch.

**Comment by rysweet** [comment]
closing as stale, please reopen I you would like to update

**Pull request closed without merging**